File: exo4_2.py
import os
import sys

def main():
    r, w = os.pipe()

    fils = os.fork()

    if fils == 0:
        os.close(w)
        os.dup2(r, sys.stdin.fileno())
        os.close(r)
        donnee = sys.stdin.read()
        print(f"\n[Fils] J'ai reçu via mon stdin redirigé : {donnee}")
        
    else:
        os.close(r)
        os.dup2(w, sys.stdout.fileno())
        os.close(w)
        print("Ceci est un message envoyé depuis le stdout du père.")
        sys.stdout.flush()
        os.close(sys.stdout.fileno())
        os.wait()

if __name__ == "__main__":
    main()

File: exo4_2.c
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>

int main () {
	int p[2];
	pipe(p);

	if (fork() == 0) {
		dup2(p[1], STDOUT_FILENO);
		close(p[0]);
		close(p[1]);

		execlp("ls", "ls", "-l", NULL);
		perror("erreur execlp");
		exit(1);
	}
	else {
		dup2(p[0], STDIN_FILENO);
		close(p[0]);
		close(p[1]);

		execlp ("wc", "wc", "-l", NULL);
		perror ("erreur execlp");
		exit(1);
	}
	return 0;
}
